src/primitive/polyhedron.rs

```diff
@@ -0,0 +1,713 @@
+        let mut best_index = 0;
+        let mut best_dot = self.vertices[best_index].position.dot(direction);
+
+        loop {
+            let previous_best_dot = best_dot;
+            let mut edge_index = self.vertices[best_index].edge;
+            let start_edge_index = edge_index;
+
+            loop {
+                let vertex_index = self.edges[edge_index].target_vertex;
+
+                let dot = self.vertices[vertex_index].position.dot(direction);
+                if dot > best_dot {
+                    best_index = vertex_index;
+                    best_dot = dot;
+                }
+
+                edge_index = self.edges[self.edges[edge_index].twin_edge].next_edge;
+                if start_edge_index == edge_index {
+                    break;
+                }
+            }
+
+            if best_dot == previous_best_dot {
+                break;
+            }
+        }
+
+        self.vertices[best_index].position.clone()
+    }
+}
+
+fn dedup_vertices<S>(vertices: Vec<Point3<S>>) -> (Vec<Point3<S>>, HashMap<usize, usize>)
+where
+    S: BaseFloat,
+{
+    let mut vs = Vec::with_capacity(vertices.len() / 2);
+    let mut dup = HashMap::default();
+    for i in 0..vertices.len() {
+        let mut found = false;
+        for j in 0..vs.len() {
+            if vs[j] == vertices[i] {
+                dup.insert(i, j);
+                found = true;
+            }
+        }
+        if !found {
+            vs.push(vertices[i]);
+        }
+    }
+    (vs, dup)
+}
+
+fn dedup_faces(
+    faces: Vec<(usize, usize, usize)>,
+    duplicates: HashMap<usize, usize>,
+) -> Vec<(usize, usize, usize)> {
+    faces
+        .iter()
+        .map(|&(a, b, c)| {
+            (
+                match duplicates.get(&a) {
+                    Some(i) => *i,
+                    None => a,
+                },
+                match duplicates.get(&b) {
+                    Some(i) => *i,
+                    None => b,
+                },
+                match duplicates.get(&c) {
+                    Some(i) => *i,
+                    None => c,
+                },
+            )
+        })
+        .collect::<Vec<_>>()
+}
+
+/// Create half edge data structure from vertices and faces
+fn build_half_edges<S>(
+    vertices: &Vec<Point3<S>>,
+    in_faces: &Vec<(usize, usize, usize)>,
+) -> (Vec<Vertex<S>>, Vec<Edge>, Vec<Face<S>>)
+where
+    S: BaseFloat,
+{
+    let mut vertices: Vec<Vertex<S>> = vertices
+        .iter()
+        .map(|v| {
+            Vertex {
+                position: v.clone(),
+                edge: 0,
+                ready: false,
+            }
+        })
+        .collect();
+    let mut edges: Vec<Edge> = vec![];
+    let mut faces: Vec<Face<S>> = vec![];
+    let mut edge_map: HashMap<(usize, usize), usize> = HashMap::default();
+    for &(a, b, c) in in_faces {
+        let face_vertices = [a, b, c];
+        println!("{}, {}, {}", a, b, c);
```

Comment:
debug `println`?

Reply:
ah, missed those. Yes, I thought I had a bug in there, turned out my test data was corrupt.